tests/kokkos-based/helpers.hpp

```diff
@@ -166,60 +195,216 @@ bool is_same_vector(
+  if (size != v2.extent(0))
+    return std::numeric_limits<RetType>::max(); // very, very different
+  const auto v1_view = KokkosKernelsSTD::Impl::mdspan_to_view(v1);
+  const auto v2_view = KokkosKernelsSTD::Impl::mdspan_to_view(v2);
+  RetType diff = static_cast<RetType>(0);
+  Kokkos::parallel_reduce(size,
+    KOKKOS_LAMBDA(const std::size_t i, RetType &diff){
+        const RetType d = scalar_abs_diff(v1_view[i], v2_view[i]);
+        if (d > diff) {
+          diff = d;
+        }
+	    }, Kokkos::Max<RetType>(diff));
+  return diff;
 }
 
-template <typename T>
-T scalar_diff(const std::complex<T> &v1, const std::complex<T> &v2) {
-  const auto dr = scalar_diff(v1.real(), v2.real());
-  const auto di = scalar_diff(v1.imag(), v2.imag());
-  return std::max(dr, di);
+template <typename ElementType1,
+          typename LayoutPolicy,
+          typename AccessorPolicy,
+          typename ElementType2>
+auto vector_abs_diff(
+    mdspan<ElementType1, extents<dynamic_extent>, LayoutPolicy, AccessorPolicy> v1,
+    const std::vector<ElementType2> &v2)
+{
+  return vector_abs_diff(v1, make_mdspan(v2));
 }
 
-template <typename T>
-KOKKOS_INLINE_FUNCTION
-T scalar_diff(const Kokkos::complex<T> &v1, const Kokkos::complex<T> &v2) {
-  const auto dr = scalar_diff(v1.real(), v2.real());
-  const auto di = scalar_diff(v1.imag(), v2.imag());
-  return dr > di ? dr : di; // can't use std::max on GPU
+template <typename ElementType1,
+          typename LayoutPolicy,
+          typename AccessorPolicy,
+          typename ElementType2>
+auto vector_abs_diff(
+    const std::vector<ElementType1> &v1,
+    mdspan<ElementType2, extents<dynamic_extent>, LayoutPolicy, AccessorPolicy> v2)
+{
+  return vector_abs_diff(v2, v1);
 }
 
-// tolerance based comparison
-// TODO: replace fixed `scalar_diff` with templated flexible norm, if needed in future
-template <typename ElementType,
+template <typename ElementType1, typename ElementType2>
+auto vector_abs_diff(
+    const std::vector<ElementType1> &v1,
+    const std::vector<ElementType2> &v2)
+{
+  return vector_abs_diff(make_mdspan(v1), make_mdspan(v2));
+}
+
+template <typename ElementType1,
+          typename LayoutPolicy1,
+          typename AccessorPolicy1,
+          typename ElementType2,
+          typename LayoutPolicy2,
+          typename AccessorPolicy2>
+auto vector_rel_diff(
+    mdspan<ElementType1, extents<dynamic_extent>, LayoutPolicy1, AccessorPolicy1> v1,
+    mdspan<ElementType2, extents<dynamic_extent>, LayoutPolicy2, AccessorPolicy2> v2)
+{
+  using RetType = decltype(scalar_abs_diff(v1[0], v2[0]));
+  const auto size = v1.extent(0);
+  if (size != v2.extent(0))
+    return std::numeric_limits<RetType>::max(); // very, very different
+
+  constexpr auto zero1 = static_cast<ElementType1>(0);
+  constexpr auto zero2 = static_cast<ElementType2>(0);
+  RetType abs_diff = vector_abs_diff(v1, v2);
+  auto v1_norm = scalar_abs_diff(zero1, v1[std::experimental::linalg::idx_abs_max(v1)]);
```

Comment:
Would you consider testing the index first, before using it to index into a vector?

Reply:
I'm sorry... Fixed on https://github.com/kokkos/stdBLAS/pull/193/commits/a945906066e35af077cbdcbf0cfdccb55ea2eeb3#diff-bd05072ed823b6605df5a6f4a81cdaa5985dd3eec497920bb8ce31cb9abaa7c4R123-R127.